### PR TITLE
Initial Commit for Nordic Connect SDK

### PR DESCRIPTION
**Pull request by dboling-ericsson**
Port of the nanoFramework to the Nordic Connect SDK. 

## Description
This ports the .NET nanoFramework to the Nordic Connect SDK. In doing so, the nanoFramework can be quickly ported to any Nordic Semiconductor controller that supports the NCS.
Specifically, support is added for the nrf9160 and the nrf52840. 
Boards supported in this initial PR are:
  Ericsson Ardesco Cellular IoT Demonstrator
  Nordic nrf9160DK (9160 Dev Kit)
  Nordic nrf52840DK (52840 Dev Kit)

## Motivation and Context
Nordic Semiconductor controllers are quite popular. Support for the NCS will enable support for a wide variety of Nordic chips. 

## How Has This Been Tested?
GPIO support has been tested on all 3 boards listed above.
I2C support has been tested on the Ardesco using the BME680 driver. I2C support should work on the other boards as board implementation details are covered in the Zephyr RTOS 

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [X] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [X] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [X] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=2447) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=2447) before we can accept your contribution.<br/><br/>:x: dboling-ericsson [sign now](https://cla.dotnetfoundation.org/nanoframework/nf-interpreter?pullRequest=2447)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nf-interpreter?pullRequest=2447) it.</sub>

**Comment by dboling-ericsson** [comment]
Sorry. I didn't know the process. I believe I've done this now.

Doug
________________________________

**Comment by josesimoes** [comment]
@dboling-ericsson that was it! Thanks. 👍🏻

**Comment by dboling-ericsson** [comment]
Now the build is breaking. Looks like binutils.Nordic.cmake has disappeared from my fork.
________________________________

**Comment by josesimoes** [comment]
> Now the build is breaking. Looks like binutils.Nordic.cmake has disappeared from my fork.

I can see that file there. Failure has to be because of some other reason.

**Comment by networkfusion** [comment]
Looks like the preset file has something invalid in it...
image

File: CMakePresets.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-    "version": 3,
+    "version": 2,
```
Comment (networkfusion) [inline CMakePresets.json:2]: I dont think this should have been bumped.

**Comment by dboling-ericsson** [comment]
> @dboling-ericsson that was it! Thanks. 👍🏻

There is a problem. Two actually.
First, the main cmakelist.txt file specifies a minimum cmake of 3.20. However, the presets files specify version 3. Looking at the docs, preset version 3 requires 3.21, not 3.20. So this is a simple bug in the cmakelist.txt file. But it leads to a bigger problem for me.
The Nordic tools pull cmake 3.20. If the cmake version is properly set, the Nordic cmake won't build the tree. I hadn't understood this issue and had 'fixed' the problem by mucking with the preset version changing it to 2. I didn't intend for this to get pushed back but it did. That apparently breaks all the other builds.
I fixed the problem by manually pulling down the latest cmake and putting it in front of the nordic tools path. However, this complicates the install. I'll have to figure out how to neatly deal with this.
There is also a third problem. In integrating the Nordic presets, I corrupted the cmakeuserpresets.TEMPLATE file. I'll push a change for that.

File: CMakePresets.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (dboling-ericsson) [reply CMakePresets.json]: Sorry. should be fixed now. As a reminder, version 3 isn't supported in CMake 3.20, so the cmake_minimum_required needs to be updated in the main cmakelists.txt file.

File: CMakePresets.json [hunk 2]
```diff
@@ -28,6 +28,20 @@
...
             }
         },
+        {
+            "name": "nordic-zephyr-preset",
+            "description": "Preset for Nordic Connect SDK",
+            "hidden": true,
+            "cacheVariables": {
+              "CMAKE_TOOLCHAIN_FILE": {
+                "type": "FILEPATH",
+                "value": "c:/ncs/toolchains/v2.0.0/cmake/NcsToolchainConfig.cmake"
```
[the remark this answers is not in the document]
Comment (dboling-ericsson) [reply CMakePresets.json:38]: I'm removing this from the cmakepresets file. Not needed as long as build started from within a Nordic command line build window. This requirement is described in the readme.

File: CMakePresets.json [hunk 3]
```diff
@@ -28,6 +28,20 @@
...
+                },
+                "RTOS": "NORDIC",
+                "RTOS_SOURCE_FOLDER": "c:/ncs/v2.0.0",
```
[the remark this answers is not in the document]
Comment (dboling-ericsson) [reply CMakePresets.json:41]: I'm removing this from the cmakepresets file. Not needed as long as build started from within a Nordic command line build window. This requirement is described in the readme.

File: CMakePresets.json [hunk 4]
```diff
@@ -28,6 +28,20 @@
...
+                "TOOLCHAIN_PREFIX": "C:/ncs/toolchains/v2.0.0/opt"
```
[the remark this answers is not in the document]
Comment (dboling-ericsson) [reply CMakePresets.json:42]: Removed. As long as the build is started within a Nordic command line build window, the make file properly finds the OS without these lines. However, if RTOS_SOURCE_FOLDER is specified in a userpresets file, the updated value will be used.

File: CMakeUserPresets.TEMPLATE.json [hunk 5]
```diff
@@ -952,6 +952,42 @@
             "displayName": "WEACT_F411CE",
             "configurePreset": "WEACT_F411CE"
         },
+        {
+            "name": "Nordic_Ardesco",
```
[the remark this answers is not in the document]
Comment (dboling-ericsson) [reply CMakeUserPresets.TEMPLATE.json:989]: This should be fixed now.

File: CMakePresets.json (hunk 1 above)
Comment (dboling-ericsson) [reply CMakePresets.json]: Fixed.

**Comment by networkfusion** [comment]
@dboling-ericsson 👏 for a sucessful CI build.

**Review by josesimoes — APPROVED** [review]
@dboling-ericsson awesome work! 👏🏻 💪🏻 
Let's have this merged into a develop branch so it's easier to do more work on this before it gets ready for merge into main.

Build on the pipeline is still to be addressed along with code style fixes and other bits.

**Pull request merged**